text/0000-deref-move.md

```diff
@@ -0,0 +1,325 @@
+
+## `IndexMove` trait
+
+One option considered in the early discussion was an `IndexMove` trait, similar
+to `DerefMove`, which would consume a collection and return a single element. I
+have not included this in this RFC because no compelling use case was presented,
+and the behaviour of `let s = v[0];` actually consuming `v` seems likely to add
+cognitive overhead to the language.
+
+Compared to other methods of moving out of a collection, `IndexMove` may provide
+slight micro-optimization in some cases, but it's not clear without benchmarks
+that this is actually meaningfully better than using, say,
+`.into_inter().nth(n)`. Given that the `IndexMove` was proposed largely from
+parallel construction from `DerefMove` without any specific use cases, it does
+not seem worthwhile to move further with it at this time.
+
+## Weakening `DerefMut` bound
+
+The `DerefMut` bound on `DerefMove` and, in particular, the `Deref` bound that
+it implies, constrain implementors somewhat. It requires that `Deref` be
+implementable, so the implementation is not allowed to overload dereferencing in
+such a way that it produces a new value each time. This would not be useful for
+non-`Copy` types, since they would be consumed, but a `Copy` type could in this
+fashion allow itself to be copied and then produce a new value on each
+derefence. This would be useful to avoid redefining methods on the outer type.
+
+This very much goes against the intent of the `Deref` traits to be restricted
+smart pointers, however. `Deref` is currently the only way for methods of one
+type to be callable on another, but it is deliberately limited in scope. For
+instance, one might be tempted to implement `Deref` on a newtype for this
+purpose, but this is considered unidiomatic. Additionally, this is a bandaid for
+the fact that newtypes would also often like to be able to inherent trait
+implementations, and inhereting methods is not sufficient for that.
+
+The additional requirement is to avoid complication of the following case:
+
+```rust
+struct RefMethod();
+
+trait Consume : Sized {
+    fn consume(self) {}
+}
+
+impl<'a> Consume for &'a mut RefMethod {}
+
+struct BadDeref<'a> (&'a mut RefMethod);
+
+impl<'a> Deref for BadDeref<'a> {
+    type Target = RefMethod;
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl<'a> DerefMove for BadDeref<'a> {
+    fn deref_move(self) -> Self::Target { self.0 }
+}
+
+fn main() {
+    let mut r = RefMethod{};
+    let b = BadDeref(&mut r);
+    b.consume();
+    b.consume();
+}
+```
+
+In this case, the first `b.consume()` is legal, but the second `b.consume()` is
+not since `b` was already moved out of. But if `BadDeref` also implements
+`DerefMut`, then the first `b.consume()` becomes illegal because `deref_mut`
+becomes preferred and `b` is immutable. By requiring that `DerefMove :
+DerefMut`, this edge case cannot be encountered: this code will always be
+illegal unless `b` is made mutable.
```

Comment:
I sort of understand why this is a bad edge case, but as was pointed out above there are useful situations where you can't feasibly implement `DerefMut`; however, this would still be a bit surprising in those cases.  I think what this edge case actually suggests is that `Box` gets it "right" mostly by accident; if it is important, the rules for `mut` bindings should be tweaked around `DerefMove` somehow (possibly that `&mut` methods on a target of `DerefMove` should require a `mut` binding, though I'm not sure how you justify that).  In any case, allowing this edge case still seems preferable to me to ruling out `DerefMove` on `!DerefMut` types.

Reply:
The behaviour of `Box` here is the same as the unique immutable reference that closures use to get around the difficulty of an immutable variable holding a mutable reference, which can be expressed in MIR but not in Rust. We could consider making `deref_mut` take `&uniq self`, if that type existed, and then this problem would not exist. Alternatively, simply dropping the notion of immutable variables would also solve this.